elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts

```diff
@@ -12,20 +12,27 @@
  * Removal or modification of this copyright notice is prohibited.
  */
 
-import { EMPTY_HASH } from './constants';
+import { DEFAULT_KEY_LENGTH, EMPTY_HASH, NodeSide } from './constants';
+import { Leaf } from './leaf';
 import { Database } from './types';
+import { parseBranch, parseLeaf, isLeaf, binaryExpansion } from './utils';
+import { Branch } from './branch';
+import { Empty } from './empty';
+import { InMemoryDB } from '../inmemory_db';
 
+type TreeNode = Branch | Leaf | Empty;
 export class SparseMerkleTree {
 	private readonly _db: Database;
-	private readonly _rootHash: Buffer;
 	private readonly _keyLength: number;
-	public constructor(db: Database, rootHash?: Buffer, keyLength = 36) {
-		this._db = db;
-		this._keyLength = keyLength;
-		this._rootHash = rootHash ?? EMPTY_HASH;
+	private _rootNode: TreeNode;
+
+	public constructor(options: { db?: Database; keyLength?: number }) {
+		this._db = options?.db ?? new InMemoryDB();
+		this._keyLength = options?.keyLength ?? DEFAULT_KEY_LENGTH;
+		this._rootNode = new Empty();
 	}
```

Comment:
1. I think we still need `rootHash` property in constructor for the scenario when it's needed to instantiate a new SMT from previous state root (from previous block header). 
2. It could be tricky to default use `InMemoryDB` if database is not provided because depending upon the user's system and the size of the SMT the system could quickly run out of memory and the library can fail. So, I'd suggest to make the db parameter non-optional so that the user knows up-front what he's doing.
```suggestion
		public constructor(db: Database, rootHash?: Buffer, keyLength?: number) {
		this._db = db;
		this._keyLength = keyLength ?? DEFAULT_KEY_LENGTH;
		if (rootHash) {
			const { leftHash, rightHash } = parseBranch(rootHash);
			this._rootNode = new Branch(leftHash, rightHash);
		}
		else {
			this._rootNode = new Empty();
		}
	}
```

Reply:
1. Yeah I actually kept that property with that in mind but completely forgot and removed it when I saw your previous comment. I will revert it back.
2. Good point, I will make it required prop for the constructor

Reply:
Okay, although for point 1, I still think we don't need `rootHash` or `get rootHash` as explicit properties of SMT since we have `rootNode`. But it should be possible to initialize using `rootHash` like in the suggestion above.

Reply:
I'm removing `rootNode` as we start with empty root node whereas the `rootHash` can have value other than empty hash. In order to retrieve the value of rootHash we need to do async call to db in the constructor which is not possible or we create a separate `init()` function